sequencing/batch_queue.go

```diff
@@ -0,0 +1,140 @@
+package sequencing
+
+import (
+	"sync"
+
+	"github.com/dgraph-io/badger/v4"
+	"github.com/rollkit/go-sequencing"
+)
+
+// BatchQueue ...
+type BatchQueue struct {
+	queue []sequencing.Batch
+	mu    sync.Mutex
+}
+
+// NewBatchQueue creates a new BatchQueue
+func NewBatchQueue() *BatchQueue {
+	return &BatchQueue{
+		queue: make([]sequencing.Batch, 0),
+	}
+}
+
+// AddBatch adds a new batch to the queue
+func (bq *BatchQueue) AddBatch(batch sequencing.Batch, db *badger.DB) error {
+	bq.mu.Lock()
+	bq.queue = append(bq.queue, batch)
+	bq.mu.Unlock()
+
+	// Get the hash and bytes of the batch
+	h, err := batch.Hash()
+	if err != nil {
+		return err
+	}
+
+	// Marshal the batch
+	batchBytes, err := batch.Marshal()
+	if err != nil {
+		return err
+	}
+
+	// Store the batch in BadgerDB
+	err = db.Update(func(txn *badger.Txn) error {
+		key := append(keyPrefixBatch, h...)
+		return txn.Set(key, batchBytes)
+	})
+	return err
+}
+
+// AddBatchToTheTop adds a new batch to the queue, at index 0
+func (bq *BatchQueue) AddBatchToTheTop(batch sequencing.Batch, db *badger.DB) error {
+	bq.mu.Lock()
+	bq.queue = append([]sequencing.Batch{batch}, bq.queue...)
+	bq.mu.Unlock()
+
+	// Get the hash and bytes of the batch
+	h, err := batch.Hash()
+	if err != nil {
+		return err
+	}
+
+	// Marshal the batch
+	batchBytes, err := batch.Marshal()
+	if err != nil {
+		return err
+	}
+
+	// Store the batch in BadgerDB
+	err = db.Update(func(txn *badger.Txn) error {
+		key := append(keyPrefixBatch, h...)
+		return txn.Set(key, batchBytes)
+	})
+	return err
+}
+
+// Next extracts a batch of transactions from the queue
+func (bq *BatchQueue) Next(db *badger.DB) (*sequencing.Batch, error) {
+	bq.mu.Lock()
+	defer bq.mu.Unlock()
+	if len(bq.queue) == 0 {
+		return &sequencing.Batch{Transactions: nil}, nil
+	}
+	batch := bq.queue[0]
+	bq.queue = bq.queue[1:]
+
+	h, err := batch.Hash()
+	if err != nil {
+		return &sequencing.Batch{Transactions: nil}, err
+	}
```

Comment:
similar comments here. if we are validating the batch before it enters the queue, then we should not be hitting an error here unless there was some sort of data corruption event. So having the checks here is still valid. 

However, before returning on error, we should also be removing the batch from the database, otherwise the database would get bloated with corrupt batches that aren't in the queue.